File: test-backend.mjs
#!/usr/bin/env node

/**
 * Backend API Test Script
 * Tests the backend server endpoints and WebSocket functionality
 */

import http from 'http';
import { io } from 'socket.io-client';

const BACKEND_URL = 'http://localhost:3000';
const SOCKET_URL = BACKEND_URL;

console.log('🧪 Testing Dual LLM Conversation System Backend\n');
console.log('='.repeat(50));

// Test 1: Health Check
async function testHealthCheck() {
  console.log('\n1️⃣ Testing Health Check Endpoint...');
  return new Promise((resolve, reject) => {
    http.get(`${BACKEND_URL}/health`, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode === 200) {
          console.log('✅ Health check passed:', data);
          resolve(true);
        } else {
          console.log('❌ Health check failed:', res.statusCode);
          reject(false);
        }
      });
    }).on('error', (err) => {
      console.log('❌ Health check error:', err.message);
      reject(false);
    });
  });
}

// Test 2: API Health Check
async function testAPIHealth() {
  console.log('\n2️⃣ Testing API Health Endpoint...');
  return new Promise((resolve, reject) => {
    http.get(`${BACKEND_URL}/api/health`, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode === 200) {
          console.log('✅ API health check passed:', data);
          resolve(true);
        } else {
          console.log('❌ API health check failed:', res.statusCode);
          reject(false);
        }
      });
    }).on('error', (err) => {
      console.log('❌ API health check error:', err.message);
      reject(false);
    });
  });
}

// Test 3: WebSocket Connection
async function testWebSocketConnection() {
  console.log('\n3️⃣ Testing WebSocket Connection...');
  return new Promise((resolve, reject) => {
    const socket = io(SOCKET_URL, {
      transports: ['websocket'],
      timeout: 5000
    });

    const timeout = setTimeout(() => {
      socket.disconnect();
      console.log('❌ WebSocket connection timeout');
      reject(false);
    }, 5000);

    socket.on('connect', () => {
      clearTimeout(timeout);
      console.log('✅ WebSocket connected successfully');
      console.log('   Socket ID:', socket.id);
      socket.disconnect();
      resolve(true);
    });

    socket.on('connect_error', (error) => {
      clearTimeout(timeout);
      console.log('❌ WebSocket connection error:', error.message);
      reject(false);
    });
  });
}

// Test 4: WebSocket Events
async function testWebSocketEvents() {
  console.log('\n4️⃣ Testing WebSocket Events...');
  return new Promise((resolve, reject) => {
    const socket = io(SOCKET_URL, {
      transports: ['websocket']
    });

    let eventsReceived = 0;
    const expectedEvents = ['conversation:state'];

    socket.on('connect', () => {
      console.log('   Connected, requesting conversation state...');
      socket.emit('conversation:getState');
    });

    socket.on('conversation:state', (data) => {
      eventsReceived++;
      console.log('✅ Received conversation:state event');
      console.log('   Data:', JSON.stringify(data, null, 2));
      socket.disconnect();
      if (eventsReceived >= expectedEvents.length) {
        resolve(true);
      }
    });

    socket.on('error', (error) => {
      console.log('❌ WebSocket error:', error);
      socket.disconnect();
      reject(false);
    });

    setTimeout(() => {
      if (eventsReceived < expectedEvents.length) {
        console.log('❌ Timeout waiting for events');
        socket.disconnect();
        reject(false);
      }
    }, 3000);
  });
}

// Run all tests
async function runTests() {
  const results = {
    passed: 0,
    failed: 0,
    tests: []
  };

  const tests = [
    { name: 'Health Check', fn: testHealthCheck },
    { name: 'API Health', fn: testAPIHealth },
    { name: 'WebSocket Connection', fn: testWebSocketConnection },
    { name: 'WebSocket Events', fn: testWebSocketEvents }
  ];

  for (const test of tests) {
    try {
      await test.fn();
      results.passed++;
      results.tests.push({ name: test.name, status: '✅ PASSED' });
    } catch (error) {
      results.failed++;
      results.tests.push({ name: test.name, status: '❌ FAILED' });
    }
  }

  console.log('\n' + '='.repeat(50));
  console.log('\n📊 Test Results Summary:');
  console.log('='.repeat(50));
  results.tests.forEach(test => {
    console.log(`${test.status} - ${test.name}`);
  });
  console.log('\n' + '='.repeat(50));
  console.log(`Total: ${results.passed + results.failed} tests`);
  console.log(`✅ Passed: ${results.passed}`);
  console.log(`❌ Failed: ${results.failed}`);
  console.log('='.repeat(50) + '\n');

  process.exit(results.failed > 0 ? 1 : 0);
}

// Check if backend is running first
http.get(`${BACKEND_URL}/health`, (res) => {
  if (res.statusCode === 200) {
    runTests().catch(err => {
      console.error('Test execution error:', err);
      process.exit(1);
    });
  } else {
    console.error('❌ Backend server is not running!');
    console.error('   Please start it with: cd backend && npm start');
    process.exit(1);
  }
}).on('error', (err) => {
  console.error('❌ Cannot connect to backend server!');
  console.error('   Error:', err.message);
  console.error('   Please start it with: cd backend && npm start');
  process.exit(1);
});

File: test_backend.py
#!/usr/bin/env python3
"""
Backend API Test Script
Tests the backend server endpoints and WebSocket functionality
"""

import requests
import socketio
import time
import sys

BACKEND_URL = 'http://localhost:3000'
SOCKET_URL = BACKEND_URL

def test_health_check():
    """Test health check endpoint"""
    print('\n1️⃣ Testing Health Check Endpoint...')
    try:
        response = requests.get(f'{BACKEND_URL}/health', timeout=5)
        if response.status_code == 200:
            print(f'✅ Health check passed: {response.json()}')
            return True
        else:
            print(f'❌ Health check failed: {response.status_code}')
            return False
    except Exception as e:
        print(f'❌ Health check error: {e}')
        return False

def test_api_health():
    """Test API health endpoint"""
    print('\n2️⃣ Testing API Health Endpoint...')
    try:
        response = requests.get(f'{BACKEND_URL}/api/health', timeout=5)
        if response.status_code == 200:
            print(f'✅ API health check passed: {response.json()}')
            return True
        else:
            print(f'❌ API health check failed: {response.status_code}')
            return False
    except Exception as e:
        print(f'❌ API health check error: {e}')
        return False

def test_websocket_connection():
    """Test WebSocket connection"""
    print('\n3️⃣ Testing WebSocket Connection...')
    try:
        sio = socketio.SimpleClient()
        sio.connect(SOCKET_URL, wait_timeout=5)
        print('✅ WebSocket connected successfully')
        print(f'   Socket ID: {sio.sid}')
        sio.disconnect()
        return True
    except Exception as e:
        print(f'❌ WebSocket connection error: {e}')
        return False

def test_websocket_events():
    """Test WebSocket events"""
    print('\n4️⃣ Testing WebSocket Events...')
    try:
        sio = socketio.SimpleClient()
        sio.connect(SOCKET_URL, wait_timeout=5)
        
        # Request conversation state
        sio.emit('conversation:getState')
        
        # Wait for response with longer timeout
        received = False
        timeout = time.time() + 5
        while time.time() < timeout:
            try:
                events = sio.receive(timeout=1)
                if events:
                    for event_name, data in events:
                        if event_name == 'conversation:state':
                            print('✅ Received conversation:state event')
                            print(f'   Data: {data}')
                            received = True
                            break
            except Exception:
                # Continue waiting
                pass
            if received:
                break
            time.sleep(0.1)
        
        sio.disconnect()
        
        if received:
            return True
        else:
            print('❌ Timeout waiting for events')
            return False
    except Exception as e:
        print(f'❌ WebSocket events error: {e}')
        return False

def main():
    print('🧪 Testing Dual LLM Conversation System Backend\n')
    print('=' * 50)
    
    # Check if backend is running
    try:
        response = requests.get(f'{BACKEND_URL}/health', timeout=2)
        if response.status_code != 200:
            print('❌ Backend server is not running!')
            print('   Please start it with: cd backend && python app.py')
            sys.exit(1)
    except Exception as e:
        print('❌ Cannot connect to backend server!')
        print(f'   Error: {e}')
        print('   Please start it with: cd backend && python app.py')
        sys.exit(1)
    
    # Run tests
    results = {
        'passed': 0,
        'failed': 0,
        'tests': []
    }
    
    tests = [
        ('Health Check', test_health_check),
        ('API Health', test_api_health),
        ('WebSocket Connection', test_websocket_connection),
        ('WebSocket Events', test_websocket_events)
    ]
    
    for test_name, test_func in tests:
        try:
            if test_func():
                results['passed'] += 1
                results['tests'].append({'name': test_name, 'status': '✅ PASSED'})
            else:
                results['failed'] += 1
                results['tests'].append({'name': test_name, 'status': '❌ FAILED'})
        except Exception as e:
            results['failed'] += 1
            results['tests'].append({'name': test_name, 'status': f'❌ FAILED ({e})'})
    
    # Print summary
    print('\n' + '=' * 50)
    print('\n📊 Test Results Summary:')
    print('=' * 50)
    for test in results['tests']:
        print(f"{test['status']} - {test['name']}")
    print('\n' + '=' * 50)
    print(f"Total: {results['passed'] + results['failed']} tests")
    print(f"✅ Passed: {results['passed']}")
    print(f"❌ Failed: {results['failed']}")
    print('=' * 50 + '\n')
    
    sys.exit(0 if results['failed'] == 0 else 1)

if __name__ == '__main__':
    main()

